src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertLinqQueryToForEachTests.cs

```diff
@@ -0,0 +1,2337 @@
+        {
+            string source = @"
+using System;
+using System.Linq;
+class Program
+{
+    static void Main()
+    {
+        var q2 =
+           [|from a in Enumerable.Range(1, 13)
+           join b in Enumerable.Range(1, 13) on 4 * a equals b
+           select a|];
+
+        foreach (var q in q2)
+        {
+            System.Console.Write(q);
+        }
+    }
+}";
+            string output = @"
+using System;
+using System.Linq;
+class Program
+{
+    static void Main()
+    {
+        System.Collections.Generic.IEnumerable<int> localFunction()
+        {
+            foreach (var a in Enumerable.Range(1, 13))
+            {
+                foreach (var b in Enumerable.Range(1, 13))
+                {
+                    if (Equals(4 * a, b))
+                    {
+                        yield return a;
+                    }
+                }
+            }
+        }
+
+        var q2 = localFunction();
+
+        foreach (var q in q2)
+        {
+            System.Console.Write(q);
+        }
+    }
+}";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact]
+        public async Task WhereClauseTest()
+        {
+            string source = @"
+using System;
+using System.Linq;
+class Program
+{
+    static void Main()
+    {
+        var nums = new int[] { 1, 2, 3, 4 };
+
+        var q2 = [|from x in nums
+                where (x > 2)
+                select x|];
+
+        string serializer = String.Empty;
+        foreach (var q in q2)
+        {
+            serializer = serializer + q + "" "";
+        }
+        System.Console.Write(serializer.Trim());
+    }
+}";
+
+            string output = @"
+using System;
+using System.Linq;
+class Program
+{
+    static void Main()
+    {
+        var nums = new int[] { 1, 2, 3, 4 };
+        System.Collections.Generic.IEnumerable<int> localFunction()
+        {
+            foreach (var x in nums)
+            {
+                if (x > 2)
+                {
+                    yield return x;
+                }
+            }
+        }
+
+        var q2 = localFunction();
+
+        string serializer = String.Empty;
+        foreach (var q in q2)
+        {
+            serializer = serializer + q + "" "";
+        }
+        System.Console.Write(serializer.Trim());
+    }
+}";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact]
+        public async Task WhereDefinedInType()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class Y
+{
+    public int Where(Func<int, bool> predicate)
+    {
+        return 45;
+    }
+}
+
+class P
+{
+    static void Main()
+    {
+        var src = new Y();
+        var query = [|from x in src
+                where x > 0
+                select x|]];
+
+        Console.Write(query);
+    }
+}";
+            //  should not provide a conversion because of the custom Where.
```

Comment:
>//  should not provide a conversion because of the custom Where. [](start = 12, length = 64)

Do we do the same for all query methods? Just curious.

Reply:
We do checks. However:
1. maybe, we should convert not bail out
2. likely, we should cover this with more test cases where each clause is overridden

[https://github.com/dotnet/roslyn/issues/25639](https://github.com/dotnet/roslyn/issues/25639)

---
In reply to: [176872402](https://github.com/dotnet/roslyn/pull/25362#discussion_r176872402) [](ancestors = 176872402)

Reply:
@jinujoseph so if we go permissive way, I think we should just allow this kind of thing and add warning so that users knows about it.